### DISCUSSION TITLE
Over-clocking and SPI maximum frequency

### DISCUSSION THREAD

**Discussion by Bodmer** [earlephilhower/arduino-pico#415] [General]
If the processor is clocked at frequencies above 125MHz then the peripheral clock (peri_clk) that feeds the SPI and UART is switched to 48MHz. This is because the peripherals attached to peri_clk are only specified to operate at 125MHz maximum. The automatic reduction of the clock to 48MHz means that the maximum SPI rate is decreased to 24MHz for any CPU frequency above 125MHz, this includes the default 133MHz setting with this core package.

To operate at SPI clock frequencies above 24MHz at CPU frequencies above 125MHz it is necessary to over-clock the peripheral clock.

It is possible to switch the SPI and UART clock back to the processor frequency, the maximum SPI rate is then half the processor frequency. This frequency switch must be done at the very start of setup() before any Serialx channel baud rates are set and before the SPI gets configured. Later the clock dividers for the SPI and UART baud rates will get correctly configured when requested.

The UART and SPI will get overclocked with the procesor and thus may not function at high over-clock speeds. I tested at 250MHz CPU clock and all seems to work fine.

Here is the code to boost the peri_clk speed:

```
// Get the processor sys_clk frequency in Hz
 uint32_t freq = clock_get_hz(clk_sys);

 // clk_peri does not have a divider, so input and output frequencies will be the same
 clock_configure(clk_peri,
                    0,
                    CLOCKS_CLK_PERI_CTRL_AUXSRC_VALUE_CLK_SYS,
                    freq,
                    freq);
```

**Comment by PJ5168** [comment]
Dear,
Excellent! Really works. Appreciate!
Best,
PJ.

**Comment by LinusHeu** [comment]
Thank you so much for this post!

> ***This is because*** the peripherals attached to peri_clk are only specified to operate at 125MHz maximum. 
> for any CPU frequency ***above*** 125MHz

Those 2 tiny details are a bit different. This "clk_peri switches to 48Mhz" issue also occurs when setting `board_build.f_cpu =  100000000L` in platformio.ini.

I think it occurs because, when this core starts, it sets clk_sys to the user-chosen/or default 133Mhz "CPU Speed" (Arduino IDE) / `board_build.f_cpu` (Platformio.ini) [here](https://github.com/earlephilhower/arduino-pico/blob/master/cores/rp2040/main.cpp#L79). 
This causes clk_peri to get switched to the USB clock, which only runs at 48MHz. I don't understand why.
You can see that switching in the screenshot here https://github.com/earlephilhower/arduino-pico/discussions/1644 Before that `set_sys_clock_khz(...)` call, clk_peri runs at 125MHz.

**Alternative workaround that runs RP2040 entirely within spec:**
Set "CPU Speed" to 125MHz or add `board_build.f_cpu = 125000000L` to platformio.ini
Then clk_sys runs at 125MHz and clk_peri stays at 125MHz as well.

**Comment by NuclearPhoenixx** [comment]
Works great, thanks. This simple change boosted the read speed of my external ADC from 600ksps to 800ksps, neat!

**Comment by gtortone** [comment]
Hi,

I'm trying to run CPU at 250 MHz (this is fine with set_sys_clock_khz) and QSPI clock at 125 MHz... is this possible ?

This works fine with pico-sdk but something weird happens with arduino-pico where QSPI clock goes at 62.5 MHz...

Thanks in advance
